include/boost/geometry/core/coordinate_promotion.hpp

```diff
@@ -0,0 +1,111 @@
+// Boost.Geometry
+
+// Copyright (c) 2021 Barend Gehrels, Amsterdam, the Netherlands.
+
+// Use, modification and distribution is subject to the Boost Software License,
+// Version 1.0. (See accompanying file LICENSE_1_0.txt or copy at
+// http://www.boost.org/LICENSE_1_0.txt)
+
+#ifndef BOOST_GEOMETRY_CORE_COORDINATE_PROMOTION_HPP
+#define BOOST_GEOMETRY_CORE_COORDINATE_PROMOTION_HPP
+
+#include <boost/geometry/core/coordinate_type.hpp>
+
+// TODO: move this to a future headerfile implementing traits for these types
+#include <boost/rational.hpp>
+#include <boost/multiprecision/cpp_bin_float.hpp>
+#include <boost/multiprecision/cpp_int.hpp>
+
+namespace boost { namespace geometry
+{
+
+namespace traits
+{
+
+// todo
```

Comment:
So this is the start of Adam's proposal - I didn't apply it everywhere, but the promotion is already here (below) and specialized for rational and multi precision. It should be worked out later, in another PR.
Two related types which should be replaced are moved in this file as well.

Reply:
Ok, we can build on top of it in a different PR. In case you wanted to jump into it I'll write what I think about this.

I think we should add more general utilities for numeric types. So to not only cover the promotion but also type traits that could be generalized to non-fundamental types. At minimum we'd need `is_integral`/`is_floating_point` type traits and a trait allowing to define floating point type corresponding to input integer type.

A name `promote_to_floating_point` seems to be ok. Another possibility is `make_floating_point` because `std::make_unsigned` and `std::make_signed` exist but I think the semantics is different so `promote` is fine. But I'd avoid past tense for sure because type traits use present tense so `make`, `remove`, `add` and therefore `promote`, `select`, `transcribe`, etc.

I'm not sure about the directory. On one hand this is important and these are traits which would make `core/` the correct directory. On the other hand type selection/promotion and type_traits are already in `util/`.

Anyway, these are my quick thoughts about it.

Reply:
Thanks, we'll come back to this.